File: yams/cpp/src/yams/mechanisms/positional/Arm.cpp
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

#include "yams/mechanisms/positional/Arm.hpp"

#include <wpi/framework/RobotBase.hpp>
#include <wpi/math/geometry/Rotation3d.hpp>
#include <wpi/math/geometry/Translation3d.hpp>
#include <wpi/simulation/BatterySim.hpp>
#include <wpi/simulation/RoboRioSim.hpp>
#include <wpi/smartdashboard/Mechanism2d.hpp>
#include <wpi/smartdashboard/MechanismLigament2d.hpp>
#include <wpi/smartdashboard/MechanismRoot2d.hpp>
#include <wpi/smartdashboard/SmartDashboard.hpp>
#include <wpi/math/system/DCMotor.hpp>
#include <wpi/util/Color.hpp>
#include <wpi/util/Color8Bit.hpp>
#include <wpi/commands2/Commands.hpp>
#include <wpi/units/angle.hpp>
#include <wpi/units/time.hpp>

#include <cmath>
#include <memory>
#include <string>

#include "yams/exceptions.hpp"
#include "yams/gearing/MechanismGearing.hpp"
#include "yams/motorcontrollers/simulation/ArmSimSupplier.hpp"

namespace yams::mechanisms::positional {

// ---- Constructor ------------------------------------------------------------

Arm::Arm(config::ArmConfig* config, motorcontrollers::SmartMotorController* smc)
    : SmartPositionalMechanism() {
  m_armConfig = config;
  m_smc = smc;
  m_subsystem = m_smc->GetConfig().GetSubsystem();

  if (!m_armConfig->GetTelemetryName().empty()) {
    m_name = m_armConfig->GetTelemetryName();
  }

  // Apply angular soft limits to the motor controller when configured.
  if (auto minA = m_armConfig->GetMinAngle()) {
    m_smc->SetMechanismLowerLimit(*minA);
  }
  if (auto maxA = m_armConfig->GetMaxAngle()) {
    m_smc->SetMechanismUpperLimit(*maxA);
  }

  // Seed the encoder from the configured starting position.
  if (auto startA = m_smc->GetConfig().GetStartingPosition()) {
    m_smc->SetEncoderPosition(*startA);
  }
  if (wpi::RobotBase::IsSimulation()) {
    // Configuration checks — throw descriptive exceptions like Java does.
    if (!m_armConfig->GetArmLength().has_value()) {
      throw exceptions::ArmConfigurationException(
          "Arm Length is empty", "Cannot create simulation.", "WithArmLength(wpi::units::meter_t)");
    }
    if (!m_armConfig->GetMinAngle().has_value()) {
      throw exceptions::ArmConfigurationException("Arm lower hard limit is empty",
                                                  "Cannot create simulation.",
                                                  "WithMinAngle(wpi::units::degree_t)");
    }
    if (!m_armConfig->GetMaxAngle().has_value()) {
      throw exceptions::ArmConfigurationException("Arm upper hard limit is empty",
                                                  "Cannot create simulation.",
                                                  "WithMaxAngle(wpi::units::degree_t)");
    }
    if (!m_smc->GetConfig().GetStartingPosition().has_value() &&
        !m_smc->GetConfig().GetExternalEncoderZeroOffset().has_value()) {
      throw exceptions::ArmConfigurationException("Arm starting angle is empty",
                                                  "Cannot create simulation.",
                                                  "smc.WithStartingPosition(wpi::units::degree_t)");
    }
    if (!m_smc->GetConfig().GetMOI()) {
      throw exceptions::ArmConfigurationException("Arm MOI is empty", "Cannot create simulation.",
                                                  "smc->GetConfig().WithMOI(length, mass)");
    }

    auto& gearingOpt = m_smc->GetConfig().GetMotorGearing();
    gearing::MechanismGearing gearing = gearingOpt.value_or(gearing::MechanismGearing::kOne);

    wpi::units::radian_t startAngle =
        m_smc->GetConfig().GetStartingPosition().value_or(wpi::units::turn_t{0});

    m_armSim.emplace(m_smc->GetDCMotor(), gearing.GetMechanismToRotorRatio(),
                     m_smc->GetConfig().GetMOI(), m_armConfig->GetArmLength().value(),
                     wpi::units::radian_t{m_armConfig->GetMinAngle().value()},
                     wpi::units::radian_t{m_armConfig->GetMaxAngle().value()}, true, startAngle,
                     std::array<double, 2>{0, 0.002 / 4096.0});

    wpi::units::second_t period = m_smc->GetConfig().GetClosedLoopControlPeriod().value_or(20_ms);
    m_smc->SetSimSupplier(std::make_shared<yams::motorcontrollers::simulation::ArmSimSupplier>(
        *m_armSim, [this]() { return m_smc->GetDutyCycle(); }, gearing, period));

    // Build Mechanism2d window.
    double armLengthM = m_armConfig->GetArmLength().value().value();
    m_mechanismWindow.emplace(armLengthM * 2.0 + 0.4, armLengthM * 2.0 + 0.4);
    m_mechanismRoot =
        m_mechanismWindow->GetRoot(m_name + "Root", armLengthM + 0.2, armLengthM + 0.2);

    wpi::units::degree_t startDeg = startAngle;
    m_mechanismLigament = m_mechanismRoot->Append<wpi::MechanismLigament2d>(
        m_name, armLengthM, startDeg, 6, m_armConfig->GetSimColor());
    m_setpointLigament = m_mechanismRoot->Append<wpi::MechanismLigament2d>(
        "Setpoint", armLengthM, startDeg, 3, wpi::util::Color8Bit{wpi::util::Color::WHITE});

    constexpr double kTickLength = 3.0 * 0.0254;  // 3 inches in metres
    m_mechanismRoot->Append<wpi::MechanismLigament2d>("MaxHard", kTickLength,
                                                      m_armConfig->GetMaxAngle().value(), 4,
                                                      wpi::util::Color8Bit{wpi::util::Color::LIME_GREEN});
    m_mechanismRoot->Append<wpi::MechanismLigament2d>("MinHard", kTickLength,
                                                      m_armConfig->GetMinAngle().value(), 4,
                                                      wpi::util::Color8Bit{wpi::util::Color::RED});

    auto smcUpperLimit = m_smc->GetConfig().GetMechanismUpperLimit();
    auto smcLowerLimit = m_smc->GetConfig().GetMechanismLowerLimit();
    if (smcUpperLimit.has_value() && smcLowerLimit.has_value()) {
      m_mechanismRoot->Append<wpi::MechanismLigament2d>(
          "MaxSoft", kTickLength, smcUpperLimit.value(), 4, wpi::util::Color8Bit{wpi::util::Color::HOT_PINK});
      m_mechanismRoot->Append<wpi::MechanismLigament2d>(
          "MinSoft", kTickLength, smcLowerLimit.value(), 4, wpi::util::Color8Bit{wpi::util::Color::YELLOW});
    }

    wpi::SmartDashboard::PutData(m_name + "/mechanism", &(*m_mechanismWindow));
  }
}

// ---- SmartMechanism overrides -----------------------------------------------

void Arm::SimIterate() {
  if (m_armSim.has_value() && m_smc->GetSimSupplier()) {
    auto* ss = m_smc->GetSimSupplier();
    ss->UpdateSim();
    m_smc->SimIterate();
    ss->StarveWatchdog();

    if (m_armConfig->GetMinAngle() && m_armSim->GetVelocity().value() < 0.0 &&
        GetAngle() < *m_armConfig->GetMinAngle()) {
      m_smc->SetEncoderPosition(*m_armConfig->GetMinAngle());
    }
    if (m_armConfig->GetMaxAngle() && m_armSim->GetVelocity().value() > 0.0 &&
        GetAngle() > *m_armConfig->GetMaxAngle()) {
      m_smc->SetEncoderPosition(*m_armConfig->GetMaxAngle());
    }
    wpi::sim::RoboRioSim::SetVInVoltage(
        wpi::sim::BatterySim::Calculate({ss->GetCurrentDrawAmps()}));
    VisualizationUpdate();
  }
}

void Arm::UpdateTelemetry() { m_smc->UpdateTelemetry(); }

void Arm::VisualizationUpdate() {
  if (m_mechanismLigament) {
    m_mechanismLigament->SetAngle(GetAngle());
  }
  if (m_setpointLigament) {
    m_setpointLigament->SetAngle(m_smc->GetMechanismPositionSetpoint().value_or(GetAngle()));
  }
}

std::string Arm::GetName() const { return m_name; }

// ---- SmartPositionalMechanism overrides -------------------------------------

wpi::cmd::Trigger Arm::Max() {
  return wpi::cmd::Trigger{
      [this] { return GetAngle() >= m_armConfig->GetMaxAngle().value_or(wpi::units::degree_t{36000}); }};
}

wpi::cmd::Trigger Arm::Min() {
  return wpi::cmd::Trigger{[this] {
    return GetAngle() <= m_armConfig->GetMinAngle().value_or(wpi::units::degree_t{-36000});
  }};
}

// ---- Arm-specific interface -------------------------------------------------

wpi::cmd::CommandPtr Arm::Run(wpi::units::degree_t angle) {
  return wpi::cmd::Run([this, angle] { SetMechanismPositionSetpoint(angle); }, {m_subsystem})
      .WithName(m_name + " Run");
}

wpi::cmd::CommandPtr Arm::Run(std::function<wpi::units::degree_t()> angle) {
  return wpi::cmd::Run([this, angle] { SetMechanismPositionSetpoint(angle()); }, {m_subsystem})
      .WithName(m_name + " Run Supplier");
}

wpi::cmd::CommandPtr Arm::RunTo(wpi::units::degree_t angle, wpi::units::degree_t tolerance) {
  wpi::cmd::Trigger near = IsNear(angle, tolerance).Debounce(wpi::units::second_t{0.1});
  return wpi::cmd::RunOnce([this, angle] { SetMechanismPositionSetpoint(angle); }, {m_subsystem})
      .AndThen(wpi::cmd::WaitUntil([near] { return near.Get(); }))
      .WithName(m_name + " RunTo");
}

wpi::cmd::CommandPtr Arm::RunTo(std::function<wpi::units::degree_t()> angle, wpi::units::degree_t tolerance) {
  wpi::units::degree_t target = angle();
  wpi::cmd::Trigger near = IsNear(target, tolerance).Debounce(wpi::units::second_t{0.1});
  return wpi::cmd::RunOnce([this, target] { SetMechanismPositionSetpoint(target); }, {m_subsystem})
      .AndThen(wpi::cmd::WaitUntil([near] { return near.Get(); }))
      .WithName(m_name + " RunTo Supplier");
}

wpi::cmd::Trigger Arm::Gte(wpi::units::degree_t angle) {
  return wpi::cmd::Trigger{[this, angle] { return GetAngle() >= angle; }};
}

wpi::cmd::Trigger Arm::Lte(wpi::units::degree_t angle) {
  return wpi::cmd::Trigger{[this, angle] { return GetAngle() <= angle; }};
}

wpi::cmd::Trigger Arm::Between(wpi::units::degree_t start, wpi::units::degree_t end) {
  return Gte(start) && (Lte(end));
}

wpi::cmd::Trigger Arm::IsNear(wpi::units::degree_t angle, wpi::units::degree_t within) {
  return wpi::cmd::Trigger{[this, angle, within] {
    return std::abs(GetAngle().value() - angle.value()) <= within.value();
  }};
}

const config::ArmConfig& Arm::GetConfig() const { return *m_armConfig; }

wpi::math::Translation3d Arm::GetRelativeMechanismPosition() const {
  if (m_mechanismLigament) {
    return wpi::math::Translation3d{wpi::units::meter_t{m_mechanismLigament->GetLength()},
                              wpi::math::Rotation3d{0_rad, 0_rad, wpi::units::radian_t{GetAngle()}}};
  }
  return wpi::math::Translation3d{};
}

void Arm::SetAngle(wpi::units::degree_t angle) { SetMechanismPositionSetpoint(angle); }

wpi::units::degree_t Arm::GetAngle() const { return m_smc->GetMechanismPosition(); }

}  // namespace yams::mechanisms::positional

File: yams/java/yams/mechanisms/positional/Arm.java
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

package yams.mechanisms.positional;

import static edu.wpi.first.units.Units.Degrees;
import static edu.wpi.first.units.Units.Inch;
import static edu.wpi.first.units.Units.Meters;
import static edu.wpi.first.units.Units.Radians;
import static edu.wpi.first.units.Units.Rotations;

import edu.wpi.first.math.filter.Debouncer.DebounceType;
import edu.wpi.first.math.geometry.Rotation3d;
import edu.wpi.first.math.geometry.Translation3d;
import edu.wpi.first.math.system.plant.DCMotor;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj.simulation.BatterySim;
import edu.wpi.first.wpilibj.simulation.RoboRioSim;
import edu.wpi.first.wpilibj.simulation.SingleJointedArmSim;
import edu.wpi.first.wpilibj.smartdashboard.Mechanism2d;
import edu.wpi.first.wpilibj.smartdashboard.MechanismLigament2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj.util.Color;
import edu.wpi.first.wpilibj.util.Color8Bit;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.Commands;
import edu.wpi.first.wpilibj2.command.button.Trigger;
import java.util.Optional;
import java.util.function.Supplier;
import yams.exceptions.ArmConfigurationException;
import yams.gearing.MechanismGearing;
import yams.mechanisms.config.ArmConfig;
import yams.mechanisms.config.MechanismPositionConfig;
import yams.mechanisms.config.MechanismPositionConfig.Plane;
import yams.motorcontrollers.SmartMotorController;
import yams.motorcontrollers.SmartMotorControllerConfig;
import yams.motorcontrollers.simulation.ArmSimSupplier;

/**
 * Arm mechanism.
 *
 * <h2>Usage Example</h2>
 * <pre>{@code
 * // Build and construct
 * SmartMotorController motor = new SparkWrapper(
 *     new SparkMax(1, MotorType.kBrushless), DCMotor.getNEO(1),
 *     new SmartMotorControllerConfig().withClosedLoopController(0.2,0,0).withStatorCurrentLimit(Amps.of(40)));
 * ArmConfig armConfig = new ArmConfig().withLength(Meters.of(0.5));
 * Arm arm = new Arm(armConfig, motor);
 *
 * // Schedule a setpoint command
 * Command moveToScore = arm.setAngle(Degrees.of(80));
 * Command holdAtZero = arm.runTo(Degrees.of(0), Degrees.of(2));
 *
 * // Bind triggers
 * arm.isNear(Degrees.of(80), Degrees.of(2)).onTrue(indexer.run());
 * arm.max().onTrue(Commands.print("Arm at max!"));
 *
 * // Call in robotPeriodic() or a subsystem's periodic():
 * arm.simIterate();
 * arm.updateTelemetry();
 * }</pre>
 */
public class Arm extends SmartPositionalMechanism
{
  /**
   * Arm config.
   */
  private final ArmConfig                     m_config;
  /**
   * Simulation for the arm.
   */
  private Optional<SingleJointedArmSim> m_sim              = Optional.empty();
  /**
   * Mechanism ligament for the setpoint.
   */
  private MechanismLigament2d           m_setpointLigament = null;

  /**
   * Constructor for the Arm mechanism.
   *
   * @param config {@link ArmConfig} to use.
   * @param smc {@link SmartMotorController} for the Arm.
   */
  public Arm(ArmConfig config, SmartMotorController smc)
  {
    this.m_config = config;
    m_smc = smc;
    SmartMotorControllerConfig smccfg = smc.getConfig();
    DCMotor          dcmotor = m_smc.getDCMotor();
    MechanismGearing gearing = m_smc.getConfig().getGearing();
    m_subsystem = smc.getConfig().getSubsystem();
    // Seed the relative encoder
    if (m_smc.getConfig().getExternalEncoder().isPresent())
    {
      m_smc.seedRelativeEncoder();
    }
    if (config.getTelemetryName().isPresent())
    {
      // TODO: Add telemetry units to config.
      m_telemetry.setupTelemetry(getName(), m_smc);
    }

    if (RobotBase.isSimulation())
    {
      if (config.getLength().isEmpty())
      {
        throw new ArmConfigurationException("Arm Length is empty", "Cannot create simulation.", "withLength(Distance)");
      }
      if (config.getLowerHardLimit().isEmpty())
      {
        throw new ArmConfigurationException("Arm lower hard limit is empty",
                                            "Cannot create simulation.",
                                            "withHardLimits(Angle,Angle)");
      }
      if (config.getUpperHardLimit().isEmpty())
      {
        throw new ArmConfigurationException("Arm upper hard limit is empty",
                                            "Cannot create simulation.",
                                            "withHardLimits(Angle,Angle)");
      }
      if (smccfg.getStartingPosition().isEmpty() && smccfg.getZeroOffset().isEmpty())
      {
        throw new ArmConfigurationException("Arm starting angle is empty",
                                            "Cannot create simulation.",
                                            "SmartMotorControllerConfig.withStartingPosition(Angle)");
      }
      if (smccfg.getStartingPosition().isPresent() &&
          (smccfg.getStartingPosition().get().lt(config.getLowerHardLimit().get()) ||
                  smccfg.getStartingPosition().get().gt(config.getUpperHardLimit().get())))
      {
        throw new ArmConfigurationException("Arm starting angle is outside hard limits",
                                            "Cannot create simulation.",
                                            "SmartMotorControllerConfig.withStartingPosition(Angle)");
      }
      m_sim = Optional.of(new SingleJointedArmSim(smc.getDCMotor(),
                                                  smccfg.getGearing().getMechanismToRotorRatio(),
                                                  smccfg.getMOI(),
                                                  config.getLength().get().in(Meters),
                                                  config.getLowerHardLimit().get().in(Radians),
                                                  config.getUpperHardLimit().get().in(Radians),
                                                  true,
                                                  smccfg.getStartingPosition().orElse(Rotations.zero()).in(Radians),
                                                  0.002 / 4096.0,
                                                  0.0));// Add noise with a std-dev of 1 tick
      m_smc.setSimSupplier(new ArmSimSupplier(m_sim.get(), m_smc));

      m_mechanismWindow = new Mechanism2d(config.getMechanismPositionConfig()
                                                .getWindowXDimension(config.getLength().get()).in(Meters),
                                          config.getMechanismPositionConfig()
                                                .getWindowYDimension(config.getLength().get()).in(Meters));
      m_mechanismRoot = m_mechanismWindow.getRoot(getName() + "Root",
                                                  config.getMechanismPositionConfig()
                                                        .getMechanismX(config.getLength().get()).in(Meters)
                                                  + config.getMechanismPositionConfig().getRelativePosition()
                                                          .orElse(new Translation3d()).getX(),
                                                  config.getMechanismPositionConfig()
                                                        .getMechanismY(config.getLength().get()).in(Meters)
                                                  + config.getMechanismPositionConfig().getRelativePosition()
                                                          .orElse(new Translation3d()).getZ()
                                                 );

      m_mechanismLigament = m_mechanismRoot.append(new MechanismLigament2d(getName(),
                                                                           config.getLength().get().in(Meters),
                                                                           smccfg.getStartingPosition()
                                                                                 .orElse(Rotations.zero()).in(Degrees),
                                                                           6,
                                                                           config.getSimColor()));
      m_setpointLigament = m_mechanismRoot.append(new MechanismLigament2d("Setpoint",
                                                                          config.getLength().get()
                                                                                .in(Meters),
              smccfg.getStartingPosition()
                                                                                .orElse(Rotations.zero())
                                                                                .in(Degrees),
                                                                          3,
                                                                          new Color8Bit(Color.kWhite)));
      m_mechanismRoot.append(new MechanismLigament2d("MaxHard",
                                                     Inch.of(3).in(Meters),
                                                     config.getUpperHardLimit().get()
                                                           .in(Degrees),
                                                     4,
                                                     new Color8Bit(Color.kLimeGreen)));
      m_mechanismRoot.append(new MechanismLigament2d("MinHard", Inch.of(3).in(Meters),
                                                     config.getLowerHardLimit().get()
                                                           .in(Degrees),
                                                     4, new Color8Bit(Color.kRed)));
      if (smccfg.getMechanismLowerLimit().isPresent() &&
          smccfg.getMechanismUpperLimit().isPresent())
      {
        m_mechanismRoot.append(new MechanismLigament2d("MaxSoft",
                                                       Inch.of(3).in(Meters),
                                                       smccfg.getMechanismUpperLimit().get()
                                                            .in(Degrees),
                                                       4,
                                                       new Color8Bit(Color.kHotPink)));
        m_mechanismRoot.append(new MechanismLigament2d("MinSoft", Inch.of(3).in(Meters),
                                                       smccfg.getMechanismLowerLimit().get()
                                                            .in(Degrees),
                                                       4, new Color8Bit(Color.kYellow)));
      }
      SmartDashboard.putData(getName() + "/mechanism", m_mechanismWindow);
      m_smc.setupSimulation();
    }
  }

  @Override
  public void updateTelemetry()
  {
//    m_telemetry.updatePosition(getAngle());
//    m_motor.getMechanismPositionSetpoint().ifPresent(m_setpoint -> m_telemetry.updateSetpoint(m_setpoint));
    m_smc.updateTelemetry();
    m_telemetry.updateLoopTime();
  }

  @Override
  public void simIterate()
  {
    if (m_sim.isPresent() && m_smc.getSimSupplier().isPresent())
    {
      m_smc.getSimSupplier().get().updateSimState();
      m_smc.simIterate();
      m_smc.getSimSupplier().get().starveUpdateSim();
      if (m_config.getLowerHardLimit().isPresent() && m_sim.get().getVelocityRadPerSec() < 0 &&
          m_smc.getMechanismPosition().lt(m_config.getLowerHardLimit().get()))
      {
        m_smc.setEncoderPosition(m_config.getLowerHardLimit().get());
      }
      if (m_config.getUpperHardLimit().isPresent() && m_sim.get().getVelocityRadPerSec() > 0 &&
          m_smc.getMechanismPosition().gt(m_config.getUpperHardLimit().get()))
      {
        m_smc.setEncoderPosition(m_config.getUpperHardLimit().get());
      }
      RoboRioSim.setVInVoltage(BatterySim.calculateDefaultBatteryLoadedVoltage(m_sim.get().getCurrentDrawAmps()));
      visualizationUpdate();
    }
  }

  /**
   * Updates the mechanism ligament with the current angle of the arm.
   *
   * @see SmartPositionalMechanism#visualizationUpdate()
   */
  @Override
  public void visualizationUpdate()
  {
    m_mechanismLigament.setAngle(getAngle().in(Degrees));
    m_setpointLigament.setAngle(m_smc.getMechanismPositionSetpoint().orElse(getAngle()).in(Degrees));
  }

  /**
   * Get the relative position of the mechanism, taking into account the relative position defined in the
   * {@link MechanismPositionConfig}.
   *
   * @return The relative position of the mechanism as a {@link Translation3d}.
   */
  @Override
  public Translation3d getRelativeMechanismPosition()
  {
    Plane movementPlane = m_config.getMechanismPositionConfig().getMovementPlane();
    Translation3d mechanismTranslation = new Translation3d(m_mechanismLigament.getLength(),
                                                           new Rotation3d(
                                                               Plane.YZ == movementPlane
                                                               ? m_mechanismLigament.getAngle()
                                                               : 0,
                                                               Plane.XZ == movementPlane
                                                               ? m_mechanismLigament.getAngle()
                                                               : 0, 0));
    if (m_config.getMechanismPositionConfig().getRelativePosition().isPresent())
    {
      return m_config.getMechanismPositionConfig().getRelativePosition().get()
                     .plus(mechanismTranslation);
    }
    return mechanismTranslation;
  }

  @Override
  public String getName()
  {
    return m_config.getTelemetryName().orElse("Arm");
  }

  /**
   * Get the {@link SmartMotorController} Mechanism Position representing the arm.
   *
   * @return Arm {@link Angle}
   */
  public Angle getAngle()
  {
    return m_smc.getMechanismPosition();
  }

  /**
   * Set the arm to the given angle.
   *
   * @param angle Arm angle to go to.
   * @return {@link Command} that sets the arm to the desired angle.
   */
  public Command setAngle(Angle angle)
  {
    return run(angle).withName(m_subsystem.getName() + " SetAngle");
  }

  /**
   * Set the arm to the given angle via a supplier.
   *
   * @param angle Supplier for the arm angle to go to.
   * @return {@link Command} that sets the arm to the desired angle.
   */
  public Command setAngle(Supplier<Angle> angle)
  {
    return run(angle).withName(m_subsystem.getName() + " SetAngle Supplier");
  }

  /**
   * Set the arm to the given angle.
   *
   * @param angle Arm angle to go to.
   * @return {@link Command} that sets the arm to the desired angle.
   */
  public Command run(Angle angle)
  {
    return Commands.run(() -> m_smc.setPosition(angle), m_subsystem).withName(m_subsystem.getName() + " SetAngle");
  }

  /**
   * Set the arm to the given angle via a supplier.
   *
   * @param angle Supplier for the arm angle to go to.
   * @return {@link Command} that sets the arm to the desired angle.
   */
  public Command run(Supplier<Angle> angle)
  {
    return Commands.run(() -> m_smc.setPosition(angle.get()), m_subsystem).withName(
        m_subsystem.getName() + " RunAngle Supplier");
  }

  /**
   * Set the arm to the given angle then end the command.
   *
   * @param angle     {@link Angle} to go to.
   * @param tolerance Tolerance {@link Angle}
   * @return {@link Command} that sets the arm to the desired angle.
   * @implNote This command will not stop. It should NOT be used when there is a default command on the Subsystem.
   */
  public Command runTo(Angle angle, Angle tolerance)
  {
    return Commands.runOnce(() -> m_smc.setPosition(angle), m_subsystem)
                   .andThen(Commands.waitUntil(isNear(angle, tolerance).debounce(0.1, DebounceType.kRising)))
                   .withName(m_subsystem.getName() + " RunTo Angle");
  }

  /**
   * Set the arm to the given angle then end the command.
   *
   * @param angle     {@link Angle} to go to.
   * @param tolerance Tolerance {@link Angle}
   * @return {@link Command} that sets the arm to the desired angle.
   * @implNote This command will stop, but the last control request to the motor controller will continue. It should NOT be used when there is a default command on the Subsystem.
   */
  public Command runTo(Supplier<Angle> angle, Angle tolerance)
  {
    return Commands.runOnce(() -> m_smc.setPosition(angle.get()), m_subsystem)
                   .andThen(Commands.waitUntil(isNear(angle.get(), tolerance).debounce(0.1, DebounceType.kRising)))
                   .withName(m_subsystem.getName() + " RunTo Angle Supplier");
  }

  /**
   * Arm is near an angle.
   *
   * @param angle  {@link Angle} to be near.
   * @param within {@link Angle} within.
   * @return {@link Trigger} on when the arm is near another angle.
   */
  public Trigger isNear(Angle angle, Angle within)
  {
    return new Trigger(() -> getAngle().isNear(angle, within));
  }

  @Override
  public Trigger max()
  {
    if (m_smc.getConfig().getMechanismUpperLimit().isPresent())
    {
      return new Trigger(gte(m_smc.getConfig().getMechanismUpperLimit().get()));
    }
    if (m_config.getUpperHardLimit().isPresent())
    {
      return gte(m_config.getUpperHardLimit().get());
    }
    throw new ArmConfigurationException("Arm upper hard and motor controller soft limit is empty",
                                        "Cannot create max trigger.",
                                        "withHardLimits(Angle,Angle)");
  }

  @Override
  public Trigger min()
  {
    if (m_smc.getConfig().getMechanismLowerLimit().isPresent())
    {
      return new Trigger(lte(m_smc.getConfig().getMechanismLowerLimit().get()));
    }
    if (m_config.getLowerHardLimit().isPresent())
    {
      return lte(m_config.getLowerHardLimit().get());
    }
    throw new ArmConfigurationException("Arm lower hard and motor controller soft limit is empty",
                                        "Cannot create min trigger.",
                                        "withHardLimits(Angle,Angle)");
  }

  /**
   * Between two angles.
   *
   * @param start Start angle.
   * @param end   End angle
   * @return {@link Trigger}
   */
  public Trigger between(Angle start, Angle end)
  {
    return gte(start).and(lte(end));
  }

  /**
   * Less than or equal to angle
   *
   * @param angle {@link Angle} to check against
   * @return {@link Trigger}
   */
  public Trigger lte(Angle angle)
  {
    return new Trigger(() -> getAngle().lte(angle));
  }

  /**
   * Greater than or equal to angle.
   *
   * @param angle Angle to check against.
   * @return {@link Trigger} for Arm.
   */
  public Trigger gte(Angle angle)
  {
    return new Trigger(() -> getAngle().gte(angle));
  }

  /**
   * Get the {@link ArmConfig} for this {@link Arm}.
   *
   * @return The {@link ArmConfig} used to configure this {@link Arm}.
   */
  public ArmConfig getArmConfig()
  {
    return m_config;
  }
}
